enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,496 @@
+workers also apply globally across all additional added workers. If a large
+amount of workers/pods is desired the user would have to re-install the cluster
+as a regular multi control-plane node cluster.
+
+- Deal with clusters that have less nodes than they had during day-1
+installation. The rationale for this being a non-goal is that it seems that
+removing the "initial" nodes of a cluster causes a lot of issues even in a
+regular, multi-node installation. So dealing with similar issues in single
+control-plane node clusters is also out of scope for this enhancement. For
+example, installing a cluster with 3 control-plane nodes and 3 workers results
+in unschedulable control-plane nodes. Scaling down the worker machine-sets from
+3 to 0 post-installation does not magically make the control-plane nodes
+schedulable, and as a result a lot of infrastructure workloads fail to schedule
+and operators degrade. User intervention is required to fix this.
+
+- Deal with expansion of the single-node control-plane by adding more control-plane nodes
+
+- Deal with expansion of clusters that have been installed before the
+implementation of this enhancement (if possible, their expansion may be
+addressed by an upgrade followed with documentation outlining the steps that
+the user has to take to enable expansion, e.g. patch the default
+`IngressController` to have node selector targetting the master pool rather
+than the worker pool).
+
+- Adjust the baremetal platform to support single control-plane node
+installations. The baremetal platform solves the "floating ingress" issue by
+using virtual IP addresses/keepalived.
+
+- Deal with users who want their ingress traffic to not go through the single
+control-plane node.
+
+- Deal with non-cloud, non-"none" platforms such as baremetal, vSphere, etc.
+
+## Proposal
+
+- Create a new `IngressTopology` API parameter.
+
+- Make sure worker node ignition files are generated even in bootstrap-in-place
+single control-plane node "none"-platform installation.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I want to install a single control-plane
+node cluster with additional worker nodes, so that my cluster will be able to
+handle larger computation demands.
+
+- As an OpenShift cluster administrator, I want to add worker nodes to my
+existing single control-plane node cluster, so that it'll be able to meet
+growing computation demands.
+
+### API Extensions
+
+Introduce a new topology field in the Ingress config CR
+(`config.openshift.io/v1/ingresses`) called `IngressTopology`.
+
+### Implementation Details/Notes/Constraints
+
+Unlike the existing `config.openshift.io/v1/infrastructures` topology fields, this
+new field will have one of these values - `ControlPlane` or `Workers`. 
+
+In addition, allow the `.spec.replicas` and `.spec.nodePlacement` parameters in
+`operator.openshift.io/v1/ingresscontrollers` CRs to be ommitted.
```

Comment:
If there are already fields that control how many ingress controller replicas there are and where they run, why do we need the new field at all? The description below only mentions default behaviors. Are those defaults different from what a user would see today?

Reply:
I've raised this concern and @staebler had [this](https://github.com/openshift/enhancements/pull/996#discussion_r799552984) to say about it

Reply:
>The description below only mentions default behaviors. Are those defaults different from what a user would see today?

See quote from enhancement:
>When the value of `IngressPlacement` is `Workers`, the defaulting behavior of
`.spec.replicas` and `.spec.nodePlacement` will be the same as it is today

Reply:
> > If there are already fields that control how many ingress controller replicas there are and where they run, why do we need the new field at all?
> 
> I've raised this concern and @staebler had [this](https://github.com/openshift/enhancements/pull/996#discussion_r799552984) to say about it

That argument explains why the new field is on the Ingress CRD, but I still don't understand why we need it at all.

> > The description below only mentions default behaviors. Are those defaults different from what a user would see today?
> 
> See quote from enhancement:
> 
> > When the value of `IngressPlacement` is `Workers`, the defaulting behavior of
> > `.spec.replicas` and `.spec.nodePlacement` will be the same as it is today

So the new field allows us to make the other fields optional and can be used to control the defaults of the other fields.

We can we not use the existing fields, and set their values using the same logic that would be used to set the value of the new field?

Reply:
>We can we not use the existing fields, and set their values using the same logic that would be used to set the value of the new field?

Setting just those fields would require the installer to create an entire IngressController CR (typically it's the Ingress operator's job to create that CR when it's missing), and this is what @staebler wants to avoid

Reply:
Ah, I missed that the fields were in different CRs. That makes much more sense now, thank you.